Mathlib/Data/Int/ModEq.lean

```diff
@@ -209,6 +218,32 @@ lemma of_mul_right (m : ℤ) : a ≡ b [ZMOD n * m] → a ≡ b [ZMOD n] :=
   mul_comm m n ▸ of_mul_left _
 #align int.modeq.of_mul_right Int.ModEq.of_mul_right
 
+/-- To cancel a common factor `c` from a `ModEq` we must divide the modulus `m` by `gcd m c`. -/
+theorem cancel_right_div_gcd (hm : 0 < m) (h : a * c ≡ b * c [ZMOD m]) : a ≡ b [ZMOD m / gcd m c] :=
+  by
+  letI d := gcd m c
+  have hmd := gcd_dvd_left m c
+  have hcd := gcd_dvd_right m c
+  rw [modEq_iff_dvd] at h⊢
+  refine' Int.dvd_of_dvd_mul_right_of_gcd_one _ _
+  -- porting note: The `show` below doesn't assign the metavariable properly, so we need an extra
+  -- `swap`
+  swap
```

Comment:
More accurately, `show` is no longer selecting the goal that matches, right?

Reply:
No it does. I suggest you try it out in gitpod. Wth the `swap`, you get a `?refine'1` metavariable in the goal. With `swap`, you get a `sorryAx` instead. I think this is a bug.

Reply:
It's doing exactly what I describe. The error is:
```
type mismatch
  this
has type
  m / ↑d ∣ c / ↑d * (b - a) : Prop
but is expected to have type
  ℤ : Type
All Messages (3)
```
`show` is supposed to be choosing a goal based on the type you give it, but it seems to be giving up and using the first one.

Reply:
mwe:
```lean
example (P Q : Prop) : P ∧ Q := by
  constructor
  show Q  -- fails
  sorry
  sorry
```

Reply:
leanprover-community/mathlib4#3305